inference-engine/src/auto_plugin/auto_plugin.cpp

```diff
@@ -24,20 +25,142 @@ namespace {
+        }
+        return {};
+    }
+
     DeviceInformation SelectDevice(const std::vector<DeviceInformation>& metaDevices) {
+        if (metaDevices.empty()) {
+            IE_THROW(NotFound) << "No available device to select in AUTO plugin";
+        }
+        if (metaDevices.size() == 1) {
+            return metaDevices.at(0);
+        }
+
+        std::vector<DeviceInformation> CPU;
+        std::vector<DeviceInformation> GPU;
+
+        for (auto& item : metaDevices) {
+            if (item.deviceName.find("CPU") == 0) {
+                CPU.push_back(item);
+                continue;
+            }
+            if (item.deviceName.find("GPU") == 0) {
+                GPU.push_back(item);
+                continue;
+            }
+        }
+
+        if (CPU.empty() && GPU.empty()) {
+            IE_THROW(NotFound) << "No available device found";
+        }
+
+        std::sort(GPU.begin(), GPU.end(), [](const DeviceInformation& a, const DeviceInformation& b)->bool{return b.deviceName < a.deviceName;});
+
+        if (!GPU.empty()) {
+            return GPU[0];
+        }
+        if (CPU.empty()) {
+            IE_THROW() << "Cannot select any device";
+        }
+        return CPU[0];
+    }
+
+    DeviceInformation SelectDevice(const InferenceEngine::CNNNetwork &network,
+                                   const std::vector<DeviceInformation>& metaDevices,
+                                   const std::vector<std::string>& optCap) {
+        if (metaDevices.empty()) {
+            IE_THROW(NotFound) << "No available device to select in AUTO plugin";
+        }
+        if (metaDevices.size() == 1) {
+            return metaDevices.at(0);
+        }
+
+        std::vector<DeviceInformation> CPU;
+        std::vector<DeviceInformation> GPU;
+
         for (auto& item : metaDevices) {
             if (item.deviceName.find("CPU") == 0) {
-              return item;
+                CPU.push_back(item);
+                continue;
+            }
+            if (item.deviceName.find("GPU") == 0) {
+                GPU.push_back(item);
+                continue;
+            }
+        }
+
+        auto networkPrecision = GetNetworkPrecision(network);
+        auto getCap = [&](std::string&& substr){
+        auto capability = std::find_if(optCap.begin(), optCap.end(),
+                                    [&](const std::string& c)->bool{ return (c.find(substr) != std::string::npos);});
+            return capability;
+        };
+
+        if (CPU.empty() && GPU.empty()) {
+            IE_THROW(NotFound) << "No available device found";
+        }
+
+        std::sort(GPU.begin(), GPU.end(), [](const DeviceInformation& a, const DeviceInformation& b)->bool{return b.deviceName < a.deviceName;});
+
+        if (!GPU.empty()) {
+            auto capability = getCap("GPU");
+            if (capability != optCap.end() && capability->find(networkPrecision) != std::string::npos) {
+                return GPU[0];
             }
         }
```

Comment:
I don't understand this logic. Looks like we need to return optimization capabilities per device, while currently it's used as a union set of optimization capabilities of AUTO sub-devices.

E.g. if GPU does not support int8, while CPU does, we still select GPU based on current logic.

Reply:
the capability is `getCap("GPU")`, so it is single device's capabilities, it's not union.

Reply:
What I see, `optCap` are represented as `BIN`, `FP16`, `INT8` etc. And you are passing CPU as `substr`.

And `getCap` is trying to match these values and I expect it cannot do this. So,
```c++
capability != optCap.end()
```
should always be `false`

Reply:
Sorry, I see, what I mean is old logic, the code has been changed, we will update.